### DISCUSSION TITLE
Client NBT Since BlockEntityClientSerializable Removal

### DISCUSSION THREAD

**Discussion by Anemony22** [FabricMC/fabric-api#1779] [Mod Dev Support] [answered]
Hi all,

I saw [here](https://github.com/FabricMC/fabric/issues/1734) that `BlockEntityClientSerializable` has been removed in the 1.18 update, and I'm curious what the new way to integrate the same functionality is now? I am using the NBT data from a block in it's model rendering. Previously the from and toClient methods worked, but I'm keen to understand these new changes.

Thank you!

-- Anemony

**Comment by warjort** [comment, marked as answer]
I assume you mean this? https://github.com/FabricMC/fabric/issues/1734
As I understand it the ````BlockEntityClientSerializable```` (BECS) is no longer necessary.

The changes made by Mojang in 1.18 (so far) are:
* the changed signature for read/writeNBT() which are now also used to handle the network packet - you get passed the NBT as a parameter
* you trigger block entity saves and updates to the client using
````
if (this.hasWorld() && !this.getWorld().isClient()) 
    ((ServerWorld) world).getChunkManager().markForUpdate(getPos());
````
There is talk of putting this logic in a helper method.

Here is an example from the testsuite:
https://github.com/FabricMC/fabric/blob/5553aff8884eca3caa823632e616873ead33324f/fabric-renderer-api-v1/src/testmod/java/net/fabricmc/fabric/test/renderer/simple/FrameBlockEntity.java
with its diff from 1.17
https://github.com/FabricMC/fabric/commit/5553aff8884eca3caa823632e616873ead33324f#diff-5a98201f7dbe103e2a01eefe44d578ee45eca7bb1ea478c4872a6bf2d4301f78

**Reply by Anemony22** [reply]
It still doesn't look like it's saving... When I try to read it from the client it returns as nothing - the default value.

You can see the class [here](https://github.com/Anemony22/Arborealis/blob/21w41a/src/main/java/com/youarethomas/arborealis/block_entities/CarvedWoodEntity.java). Have I missed something obvious?

**Reply by warjort** [reply]
It looks like that BlockEntity in the testsuite is misleading and probably wrong?

The way it works now is that you need to implement a new method to create the network packet.
By default this method returns null which is why you are not seeing anything happening (the only thing sent to the client will be the container, e.g. the itemstacks for a chest)

There is a helper method to make this easy for simple implementations:
````
    @Override
    public BlockEntityUpdateS2CPacket toUpdatePacket() {
        return BlockEntityUpdateS2CPacket.create(this);
    }
````
But even this is not enough, because this helper calls back into another method in your BlockEntity.
By default this returns empty NBT so you need to override the method to tell it to fill in the data.
````
    @Override
    public NbtCompound toInitialChunkDataNbt() {
        return this.createNbt();
    }
````

The yarn mappings for 21w41a kind of explain what is going on:
https://maven.fabricmc.net/docs/yarn-21w41a+build.9/net/minecraft/block/entity/BlockEntity.html#toUpdatePacket()

This will appear in the source for minecraft if you run ````gradlew gensources````

**Reply by Anemony22** [reply]
That's what did it! Really appreciate your time and assistance - thank you.

**Comment by warjort** [comment]
By the way, you still have the default mixin from the fabric example mod:
https://github.com/Anemony22/Arborealis/blob/21w41a/src/main/java/com/youarethomas/arborealis/mixin/ArborealisMixin.java

If you just are keeping it around for reference, I would suggest removing it from your config:
https://github.com/Anemony22/Arborealis/blob/21w41a/src/main/resources/aborealis.mixins.json

Or since this is the only mixin, you can remove the mixins config from your fabric.mod.json
